packages/babel-plugin-transform-react-display-name/src/index.js

```diff
@@ -52,59 +52,162 @@ export default declare(api => {
+      return false;
+    }
+
+    return true;
+  }
+
+  function displayNameFromFilename(filename) {
+    if (!filename) {
+      return null;
+    }
+
+    let displayName = path.basename(filename, path.extname(filename));
+
+    // ./{module name}/index.js
+    if (displayName === "index") {
+      displayName = path.basename(path.dirname(filename));
+    }
+
+    return displayName;
+  }
+
+  function displayNameFromAncestors(path) {
+    let id;
+
+    // crawl up the ancestry looking for possible candidates for displayName inference
+    path.find(function(path) {
+      if (path.isAssignmentExpression()) {
+        id = path.node.left;
+      } else if (path.isObjectProperty()) {
+        id = path.node.key;
+      } else if (path.isVariableDeclarator()) {
+        id = path.node.id;
+      } else if (path.isStatement()) {
+        // we've hit a statement, we should stop crawling up
+        return true;
+      }
+
+      // we've got an id! no need to continue
+      if (id) return true;
+    });
+
+    // ensure that we have an identifier we can inherit from
+    if (!id) return;
+
+    // foo.bar -> bar
+    if (t.isMemberExpression(id)) {
+      id = id.property;
+    }
+
+    // identifiers are the only thing we can reliably get a name from
+    if (t.isIdentifier(id)) {
+      return id.name;
+    }
+
+    return null;
+  }
+
+  function transformLazyReference(idPath, state) {
+    const callPath = idPath.parentPath;
+    if (!callPath.isCallExpression() || idPath.key !== "callee") {
+      return;
+    }
+    if (
+      callPath.parentPath.parentPath &&
+      callPath.parentPath.parentPath.isSequenceExpression()
+    ) {
+      return;
+    }
+    let displayName;
+    if (
+      callPath.parentPath.isExportDefaultDeclaration() &&
+      callPath.key === "declaration"
+    ) {
+      displayName = displayNameFromFilename(state.filename);
+    } else {
+      displayName = displayNameFromAncestors(callPath);
+    }
+    if (!displayName) {
+      let maybeImportCall, maybeImportFilename;
+      if (
+        callPath.node.arguments.length === 1 &&
+        t.isArrowFunctionExpression(callPath.node.arguments[0])
+      ) {
+        maybeImportCall = callPath.node.arguments[0].body;
+      }
+      if (
+        maybeImportCall &&
+        t.isCallExpression(maybeImportCall) &&
+        t.isImport(maybeImportCall.callee) &&
+        maybeImportCall.arguments.length === 1
+      ) {
+        maybeImportFilename = maybeImportCall.arguments[0];
+      }
+      if (maybeImportFilename && t.isStringLiteral(maybeImportFilename)) {
+        displayName = displayNameFromFilename(maybeImportFilename.value);
+      }
+    }
+    if (!displayName) {
+      return;
+    }
+    const { scope } = callPath;
+    const id = scope.generateDeclaredUidIdentifier(displayName);
+    scope.push({ id });
+    callPath.replaceWith(
+      t.sequenceExpression([
+        t.assignmentExpression("=", id, callPath.node),
```

Comment:
The `id` needs to be cloned to prevent duplicate nodes in AST. For the background, see #7149 and #10163.